### PR TITLE
Tweak to Throw Mode -- "AirDrop" mode

### PR DESCRIPTION
**Pull request by Chambana**

### CODE REVIEW

**Comment by Chambana** [comment]
Small tweak to Throw Mode.  The intent is to let you just drop the copter instead of throwing it upward.

**Comment by magicrub** [comment]
Please rebase on master. There are conflicts.

**Comment by rmackay9** [comment]
I think we should name the new parameter something like "THROW_TYPE" or something else that starts with "THROW_" so that users can find it more easily.  It could have a value of 0 = upward throw, 1 = drop.
Normally we ask that the commits start with the vehicle or library name, i.e. "Copter:".
On a more general note, I think a lot of people are looking for this enhancements.

File: ArduCopter/control_throw.cpp [hunk 1]
```diff
@@ -185,7 +194,14 @@ bool Copter::throw_detected()
     bool high_speed = inertial_nav.get_velocity().length() > 500.0f;
 
     // check for upwards trajectory
-    bool gaining_height = inertial_nav.get_velocity().z > 50.0f;
+    //bool gaining_height = inertial_nav.get_velocity().z > 50.0f;
+    bool changing_height;
+    if (g.throw_mode_airdrop_enable==1) {
+        changing_height = inertial_nav.get_velocity().z < -50.0f;
+    }
```
Comment (rmackay9) [inline ArduCopter/control_throw.cpp]: Small formatting issue - normally it should be of the format "} else {"

**Comment by Chambana** [comment]
@magicrub @rmackay9  Sure, good points.  I'll take another crack at it.

**Comment by rmackay9** [comment]
@Chambana, while you're at it.. I was wondering.. how about adding another parameter called "THROW_NEXT_MODE" that can be set to any flight mode's number (although really people would most likely only want to use Guided or Auto) and then after the throw is completed and the vehicle has stabilized it's position, it attempts to switch to this flight mode?
No pressure of course..your original change alone would be good to get into master.

File: ArduCopter/control_throw.cpp [hunk 2]
```diff
@@ -185,7 +194,14 @@ bool Copter::throw_detected()
     bool high_speed = inertial_nav.get_velocity().length() > 500.0f;
 
     // check for upwards trajectory
-    bool gaining_height = inertial_nav.get_velocity().z > 50.0f;
+    //bool gaining_height = inertial_nav.get_velocity().z > 50.0f;
```
Comment (priseborough) [inline ArduCopter/control_throw.cpp]: Don't forget to update the comment

File: ArduCopter/control_throw.cpp [hunk 3]
```diff
@@ -194,7 +210,9 @@ bool Copter::throw_detected()
     bool no_throw_action = ins.get_accel().length() < 1.0f * GRAVITY_MSS;
 
     // High velocity or free-fall combined with incresing height indicate a possible throw release
```
Comment (priseborough) [inline ArduCopter/control_throw.cpp]: Update the comment

Comment (priseborough) [inline ArduCopter/control_throw.cpp]: This comment line needs to be updated or removed

File: ArduCopter/control_throw.cpp [hunk 4]
```diff
@@ -184,17 +193,27 @@ bool Copter::throw_detected()
...
 
     // Check the vertical acceleraton is greater than 0.25g
     bool free_falling = ahrs.get_accel_ef().z > -0.25 * GRAVITY_MSS;
 
     // Check if the accel length is < 1.0g indicating that any throw action is complete and the copter has been released
     bool no_throw_action = ins.get_accel().length() < 1.0f * GRAVITY_MSS;
 
-    // High velocity or free-fall combined with incresing height indicate a possible throw release
-    bool possible_throw_detected = (free_falling || high_speed) && gaining_height && no_throw_action;
+    // High velocity or free-fall combined with increasing height indicate a possible throw release
```
Comment (priseborough) [inline ArduCopter/control_throw.cpp]: High velocity or free-fall combined with change in height indicate a possible air-drop or throw release

**Comment by Chambana** [comment]
@rmackay9 I'll give THROW_NEXT_MODE a shot, good idea.  

I've also been thinking about adding something like a THROW_STABILIZE_HEIGHT, allowing the copter to deliberately freefall until the specified throw_stabilize_height is reached.  Otherwise, if you air dropped from a very high altitude (e.g. from a cliff, balloon, bigger drone, etc), the copter would almost immediately stabilize at high altitude and then have to descend at whatever you set WPNAV_SPEED_DN to.

**Comment by magicrub** [comment]
would THROW_STABILIZE_HEIGHT be a negative height relative to home?

**Comment by Chambana** [comment]
@magicrub Hmmm, good question.  I was thinking of using the LIDAR to get height above ground, but the platform is going to be flopping around in freefall and the LIDAR will be useless!  If the copter is dropping from another platform, then perhaps you'd want to use some offset relative to the drop height (e.g. fall 300m below drop altitude before stabilizing).  Otherwise, if you have terrain data loaded, I suppose you could use terrain data to better calculate a height above ground to stabilize at.  
Using negative height relative to Home would work if Home were set at the drop height (e.g. throwing off a cliff), but for the scenario where you're dropping from another airframe then I'm not sure. In that case, Home altitude would be set at first GPS lock (e.g. while on the ground), in which case using negative height relative to home would be dicey.  I'll have to think on this some more and thanks for chiming in.

**Comment by magicrub** [comment]
> (e.g. from a cliff, balloon, bigger drone, etc)
> ..I was thinking of using the LIDAR to get height above ground
for starters, you'll be tumbling while falling so a rangefinder is useless. Second, falling from all of the 3 you listed are potentially extremely high altitudes. Even if you were falling level-ish and the lidar can see the ground, it'll be a very hard maneuver to stop a terminal-velocity drone in a short distance. Lidar is a bad choice to rely on for this.

as far as the free-fall distance, it can just be a relative drop of how long it's been in free-fall. No need to reference it to home, right?

**Comment by Chambana** [comment]
Yes, I believe that's right -- No need to reference it to home and I think the relative from drop height should work and is easiest.  

I may hit you up in Gitter to help walk me through this, but what I'm wondering is, is there any way to get height-above-ground in this situation (where your airframe is flopping around in freefall)?  You could use a combination of GNSS-derived alt and terrain data, no?  You can't use baro because that's zero'd at Home (e.g. launch site) and you might be dropping over a feature with significantly different height (e.g. mountain), please correct me if I'm wrong on that!

**Comment by magicrub** [comment]
it's just a delta height. Reference should not matter.

**Comment by rmackay9** [comment]
i think once you've got the throw-next-mode feature in there, more complex responses can be handled as part of a mission.  So instead of adding a THROW_STABILIZE_HEIGHT you could instead have a mission that gets the copter down in a controlled manner to the desired height.  If we have a problem descending at a decent speed in auto, we could address that perhaps by improving the do-set-speed mission command to accept setting the vertical speed (currently it only accepts setting the horizontal speed I think).

**Pull request closed without merging**